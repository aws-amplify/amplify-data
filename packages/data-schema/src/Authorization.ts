import type { UnionToIntersection } from '@aws-amplify/data-schema-types';

const __data = Symbol('data');

/**
 * All possible providers.
 *
 * This list should not be used if you need to restrict available providers
 * according to an auth strategcy. E.g., `public` auth can only be facilitated
 * by `apiKey` and `iam` providers.
 */
export const Providers = [
  'apiKey',
  'iam',
  'userPools',
  'oidc',
  'function',
] as const;
export type Provider = (typeof Providers)[number];

/**
 * The subset of auth providers that can facilitate `public` auth.
 */
export const PublicProviders = ['apiKey', 'iam'] as const;
export type PublicProvider = (typeof PublicProviders)[number];

/**
 * The subset of auth providers that can facilitate `private` auth.
 */
export const PrivateProviders = ['userPools', 'oidc', 'iam'] as const;
export type PrivateProvider = (typeof PrivateProviders)[number];

/**
 * The subset of auth providers that can facilitate `owner` auth.
 */
export const OwnerProviders = ['userPools', 'oidc'] as const;
export type OwnerProviders = (typeof OwnerProviders)[number];

/**
 * The subset of auth providers that can facilitate `group` auth.
 */
export const GroupProviders = ['userPools', 'oidc'] as const;
export type GroupProvider = (typeof GroupProviders)[number];

/**
 * The subset of auth providers that can facilitate `custom` auth.
 */
export const CustomProviders = ['function'] as const;
export type CustomProvider = (typeof CustomProviders)[number];

export const Strategies = [
  'public',
  'private',
  'owner',
  'groups',
  'custom',
] as const;
export type Strategy = (typeof Strategies)[number];

/**
 * The operations that can be performed against an API.
 */
export const Operations = [
  'create',
  'update',
  'delete',
  'read',
  'get',
  'list',
  'sync',
  'listen',
  'search',
] as const;
export type Operation = (typeof Operations)[number];

export type Authorization<
  AuthStrategy extends Strategy,
  AuthField extends string | undefined,
  AuthFieldPlurality extends boolean,
> = {
  [__data]: {
    strategy?: AuthStrategy;
    provider?: Provider;
    operations?: Operation[];
    groupOrOwnerField?: AuthField;
    groups?: string[];
    multiOwner: AuthFieldPlurality;
    identityClaim?: string;
    groupClaim?: string;
  };
};

export type OwnerField = object;

type BuilderMethods<T extends object> = {
  [K in keyof T as T[K] extends (...args: any) => any ? K : never]: T[K];
};

/**
 * Creates a shallow copy of an object with an individual field pruned away.
 *
 * @param original The original object to prune.
 * @param without The field to prune.
 * @returns The pruned object.
 */
function omit<T extends object, O extends string>(
  original: T,
  without: O,
): Omit<T, O> {
  const pruned = { ...original };
  delete (pruned as any)[without];
  return pruned;
}

function to<SELF extends Authorization<any, any, any>>(
  this: SELF,
  operations: Operation[],
) {
  (this as any)[__data].operations = operations;
  return omit(this, 'to');
}

function inField<
  SELF extends Authorization<any, any, any>,
  Field extends string,
>(this: SELF, field: Field) {
  this[__data].groupOrOwnerField = field;
  const built = omit(this, 'inField');

  return built as unknown as BuilderMethods<typeof built> &
    Authorization<
      SELF[typeof __data]['strategy'],
      Field,
      SELF[typeof __data]['multiOwner']
    >;
}

/**
 * Specifies a property of the identity JWT to use in place of `sub::username`
 * as the value to match against the owner field for authorization.
 *
 * @param this Authorization object to operate against.
 * @param property A property of identity JWT.
 * @returns A copy of the Authorization object with the claim attached.
 */
function identityClaim<SELF extends Authorization<any, any, any>>(
  this: SELF,
  property: string,
) {
  this[__data].identityClaim = property;
  return omit(this, 'identityClaim');
}

function withClaimIn<SELF extends Authorization<any, any, any>>(
  this: SELF,
  property: string,
) {
  this[__data].groupClaim = property;
  return omit(this, 'withClaimIn');
}

function validateProvider(
  needle: Provider | undefined,
  haystack: readonly Provider[],
) {
  if (needle && !haystack.includes(needle)) {
    throw new Error(`Invalid provider (${needle}) given!`);
  }
}

function authData<
  Strat extends Strategy = 'public',
  Field extends string | undefined = undefined,
  isMulti extends boolean = false,
  Builders extends object = object,
>(
  defaults: Partial<Authorization<Strat, Field, isMulti>[typeof __data]>,
  builderMethods: Builders,
): Authorization<Strat, Field, isMulti> & Builders {
  return {
    [__data]: {
      strategy: 'public',
      provider: undefined,
      operations: undefined,
      groupOrOwnerField: undefined,
      multiOwner: false,
      identityClaim: undefined,
      groups: undefined,
      ...defaults,
    } as any,
    ...builderMethods,
  };
}

/**
 * Defines an authorization rule for your data models and fields. First choose an authorization strategy (`public`,
 * `private`, `owner`, `group`, or `custom`), then choose an auth provider (`apiKey`, `iam`, `userPools`, `oidc`, or `function`)
 * and optionally use `.to(...)` to specify the operations that can be performed against your data models and fields.
 */
export const allow = {
  /**
   * Authorize unauthenticated users. By default, `.public()` uses an API key based authorization. You can additionally
   * use `.public('iam')` to use IAM based authorization for unauthenticated users.
   * @param provider the authentication provider - supports "apiKey" or "iam" as valid providers
   * @returns an authorization rule for unauthenticated users
   */
  public(provider?: PublicProvider) {
    validateProvider(provider, PublicProviders);
    return authData(
      {
        strategy: 'public',
        provider,
      },
      {
        to,
      },
    );
  },

  /**
   * Authorize authenticated users. By default, `.private()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.private("iam")` or `.private("oidc")` to use IAM or OIDC based authorization for authenticated users.
   * @param provider the authentication provider - supports "userPools", "iam", or "oidc"
   * @returns an authorization rule for authenticated users
   */
  private(provider?: PrivateProvider) {
    validateProvider(provider, PrivateProviders);
    return authData(
      {
        strategy: 'private',
        provider,
      },
      {
        to,
      },
    );
  },

  /**
   * Authorize access on a per-user (owner) basis. By setting owner-based authorization, a new `owner: a.string()`
   * field will be added to the model to store which user "owns" the item. Upon item creation, the "owner field" is
   * auto-populated with the authenticated user's information. You can which field should be used as the owner field
   * by chaining the * `.inField(...)` method.
   *
   * By default, `.owner()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.owner("oidc")` to use OIDC based authentication to designate the owner.
   *
   * To change the specific claim that should be used as the user identifier within the owner field, chain the
   * `.identityClaim(...)` method.
   *
   * @param provider the authentication provider - supports "userPools", "iam", or "oidc"
   * @returns an authorization rule for authenticated users
   */
  owner(provider?: OwnerProviders) {
    validateProvider(provider, OwnerProviders);
    return authData(
      {
        strategy: 'owner',
        provider,
        groupOrOwnerField: 'owner',
      },
      {
        to,
        inField,
        identityClaim,
      },
    );
  },

  /**
   * Authorize access for multi-user / multi-owner access. By setting multi-owner-based authorization, a new `owners: a.string().array()`
   * field will be added to the model to store which users "own" the item. Upon item creation, the "owners field" is
   * auto-populated with the authenticated user's information. To grant other users access to the item, append their user identifier into the `owners` array.
   * You can which field should be used as the owners field by chaining the * `.inField(...)` method.
   *
   * By default, `.multipleOwners()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.multipleOwners("oidc")` to use OIDC based authentication to designate the owner.
   *
   * To change the specific claim that should be used as the user identifier within the owners field, chain the
   * `.identityClaim(...)` method.
   *
   * @param provider the authentication provider - supports "userPools", "iam", or "oidc"
   * @returns an authorization rule for authenticated users
   */
  multipleOwners(provider?: OwnerProviders) {
    validateProvider(provider, OwnerProviders);
    return authData(
      {
        strategy: 'owner',
        multiOwner: true,
        provider,
        groupOrOwnerField: 'owner',
      },
      {
        to,
        inField,
        identityClaim,
      },
    );
  },

  /**
   * Authorize a specific user group. Provide the name of the specific user group to have access.
   *
   * By default, `.specificGroup()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.specificGroup("group-name", "oidc")` to use OIDC based authentication to designate the user group.
   *
   * To change the specific claim that should be used as the user group identifier, chain the
   * `.withClaimIn(...)` method.
   * @param group the name of the group to authorize
   * @param provider the authentication provider - supports "userPools" or "oidc"
   * @returns an authorization rule to grant access by a specific group
   */
  specificGroup(group: string, provider?: GroupProvider) {
    return authData(
      {
        strategy: 'groups',
        provider,
        groups: [group],
      },
      {
        to,
        withClaimIn,
      },
    );
  },

  /**
   * Authorize multiple specific user groups. Provide the names of the specific user groups to have access.
   *
   * By default, `.specificGroups()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.specificGroups(["group-a", "group-b"], "oidc")` to use OIDC based authentication to designate the user group.
   *
   * To change the specific claim that should be used as the user group identifier, chain the
   * `.withClaimIn(...)` method.
   * @param groups the names of the group to authorize defined as an array
   * @param provider the authentication provider - supports "userPools" or "oidc"
   * @returns an authorization rule to grant access by a specific group
   */
  specificGroups(groups: string[], provider?: GroupProvider) {
    return authData(
      {
        strategy: 'groups',
        provider,
        groups,
      },
      {
        to,
        withClaimIn,
      },
    );
  },

  /**
   * Authorize if a user is part of a group defined in a data model field.
   *
   * By default, `.groupDefinedIn()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.groupDefinedIn("field-name", "oidc")` to use OIDC based authentication to designate the user group.
   *
   * To change the specific claim that should be used as the user group identifier within the groups field, chain the
   * `.withClaimIn(...)` method.
   * @param groupsField the field that should store the authorized user group information
   * @param provider the authentication provider - supports "userPools" or "oidc"
   * @returns an authorization rule to grant access by a specific group
   */
  groupDefinedIn<T extends string>(groupsField: T, provider?: GroupProvider) {
    return authData(
      {
        strategy: 'groups',
        provider,
        groupOrOwnerField: groupsField,
      },
      {
        to,
      },
    );
  },

  /**
   * Authorize if a user is part of a one of the groups defined in a data model field.
   *
   * By default, `.groupsDefinedIn()` uses an Amazon Cognito user pool based authorization. You can additionally
   * use `.groupsDefinedIn("field-name", "oidc")` to use OIDC based authentication to designate the user group.
   *
   * To change the specific claim that should be used as the user group identifier within the groups field, chain the
   * `.withClaimIn(...)` method.
   * @param groupsField the field that should store the list of authorized user groups
   * @param provider the authentication provider - supports "userPools" or "oidc"
   * @returns an authorization rule to grant access by a specific group
   */
  groupsDefinedIn<T extends string>(groupsField: T, provider?: GroupProvider) {
    return authData(
      {
        strategy: 'groups',
        provider,
        groupOrOwnerField: groupsField,
        multiOwner: true,
      },
      {
        to,
      },
    );
  },

  custom(provider?: CustomProvider) {
    return authData(
      {
        strategy: 'custom',
        provider,
      },
      {
        to,
      },
    );
  },
} as const;

/**
 * Turns the type from a list of `Authorization` rules like this:
 *
 * ```typescript
 * [
 *  allow.public(),
 *  allow.owner().inField('otherfield'),
 *  allow.multipleOwners().inField('editors')
 * ]
 * ```
 *
 * Into a union of the possible `fieldname: type` auth objects like this:
 *
 * ```typescript
 * {
 *  owner?: string | undefined;
 * } | {
 *  otherfield?: string | undefined;
 * } | {
 *  editors?: string[] | undefined;
 * }
 * ```
 */
export type ImpliedAuthField<T extends Authorization<any, any, any>> =
  T extends Authorization<infer _Strat, infer Field, infer isMulti>
    ? Field extends undefined
      ? never
      : Field extends string
      ? isMulti extends true
        ? { [K in Field]?: string[] }
        : { [K in Field]?: string }
      : never
    : never;

/**
 * Turns the type from a list of `Authorization` rules like this:
 *
 * ```typescript
 * [
 *  allow.public(),
 *  allow.owner().inField('otherfield'),
 *  allow.multipleOwners().inField('editors')
 * ]
 * ```
 *
 * Into an object type that includes all auth fields like this:
 *
 * ```typescript
 * {
 *  owner?: string | undefined;
 *  otherfield?: string | undefined;
 *  editors?: string[] | undefined;
 * }
 * ```
 */
export type ImpliedAuthFields<T extends Authorization<any, any, any>> =
  ImpliedAuthField<T> extends never
    ? never
    : UnionToIntersection<ImpliedAuthField<T>>;

export const accessData = <T extends Authorization<any, any, any>>(
  authorization: T,
) => authorization[__data];
