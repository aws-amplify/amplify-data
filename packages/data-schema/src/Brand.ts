const brand = Symbol('brand');
type brand = typeof brand;

export type Brand<T, TBrand extends string> = T & {
  [brand]: TBrand;
};

export function brandTarget<T extends object, TBrand extends string>(
  target: T,
  brandName: TBrand,
): Brand<T, TBrand> {
  return { ...target, [brand]: brandName };
}
