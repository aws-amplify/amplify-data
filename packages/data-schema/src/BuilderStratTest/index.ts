export {
  RefType as RefDataBranding,
  ref as refBranding,
  getData as getDataBranding,
} from './RefDataBranding';
export {
  RefType as RefDataKBranding,
  ref as refKBranding,
  getData as getDataKBranding,
} from './RefDataClassBranding';
export {
  RefType as RefInternalCasting,
  InternalRef as InternalRefInternalCasting,
  ref as refInternalCasting,
} from './RefInternalCasting';
