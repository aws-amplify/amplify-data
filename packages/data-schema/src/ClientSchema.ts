import { type __modelMeta__ } from '@aws-amplify/data-schema-types';
import { type GenericModelSchema, type RDSModelSchema } from './ModelSchema';

// MappedTypes
import type { ResolveSchema, SchemaTypes } from './MappedTypes/ResolveSchema';
import type {
  CreateImplicitModelsFromRelations,
  ResolveFieldProperties,
  ResolveStaticFieldProperties,
} from './MappedTypes/ResolveFieldProperties';
import type {
  ModelIdentifier,
  ModelSecondaryIndexes,
  RelationalMetadata,
} from './MappedTypes/ModelMetadata';
import type {
  ExtractNonModelTypes,
  NonModelTypesShape,
} from './MappedTypes/ExtractNonModelTypes';
import {
  ResolveCustomOperations,
  CustomOperationHandlerTypes,
} from './MappedTypes/CustomOperations';

export type ClientSchema<Schema extends GenericModelSchema<any>> =
  InternalClientSchema<Schema>;

/**
 * Types for unwrapping generic type args into client-consumable types
 *
 * @typeParam Schema - Data schema builder model type
 *
 * The following params are used solely as variables in order to simplify mapped type usage.
 * They should not receive external type args.
 *
 * @internal @typeParam NonModelTypes - Custom Types, Enums, and Custom Operations
 * @internal @typeParam ResolvedSchema - Resolve the schema types used by other generics
 * @internal @typeParam ImplicitModels - The implicit models created to represent relationships
 * @internal @typeParam ImplicitModelsIdentifierMeta - The implicite model identifiers derived from ImplicitModels
 * @internal @typeParam ResolvedFields - Resolved client-facing types used for CRUDL response shapes
 * @internal @typeParam SecondaryIndexes - Map of model secondary index metadata
 */
type InternalClientSchema<
  Schema extends GenericModelSchema<any>,
  NonModelTypes extends NonModelTypesShape = ExtractNonModelTypes<Schema>,
  ResolvedSchema = ResolveSchema<Schema>,
  ImplicitModels = Schema extends RDSModelSchema<any, any>
    ? object
    : CreateImplicitModelsFromRelations<ResolvedSchema>,
  ImplicitModelsIdentifierMeta = {
    [ImplicitModel in keyof ImplicitModels]: {
      identifier: 'id';
    };
  },
  ResolvedFields extends Record<
    string,
    unknown
  > = Schema extends RDSModelSchema<any, any>
    ? ResolveStaticFieldProperties<Schema, NonModelTypes, object>
    : ResolveFieldProperties<Schema, NonModelTypes, ImplicitModels>,
> = InternalClientSchemaEntries<Schema, NonModelTypes, ResolvedFields> & {
  [__modelMeta__]: InternalClientSchemaMetadata<
    Schema,
    NonModelTypes,
    ResolvedSchema,
    ImplicitModelsIdentifierMeta,
    ResolvedFields
  >;
};

/**
 * Types for expanding into client-consumable types
 *
 * @typeParam Schema - Data schema builder model type
 * @typeParam NonModelTypes - Custom Types, Enums, and Custom Operations
 * @typeParam ResolvedFields - Resolved client-facing types used for CRUDL response shapes
 */
type InternalClientSchemaEntries<
  Schema extends GenericModelSchema<any>,
  NonModelTypes extends NonModelTypesShape,
  ResolvedFields extends Record<string, unknown>,
> = CustomOperationHandlerTypes<
  ResolveCustomOperations<
    Schema,
    ResolvedFields,
    NonModelTypes
  >['customOperations']
> &
  ResolvedFields;

/**
 * Types for expanding into client-consumable metadata types
 *
 * @typeParam Schema - Data schema builder model type
 *
 * The following params are used solely as variables in order to simplify mapped type usage.
 * They should not receive external type args.
 *
 * @internal @typeParam NonModelTypes - Custom Types, Enums, and Custom Operations
 * @internal @typeParam ResolvedSchema - Resolve the schema types used by other generics
 * @internal @typeParam ImplicitModelsIdentifierMeta - The implicite model identifiers derived from ImplicitModels
 * @internal @typeParam ResolvedFields - Resolved client-facing types used for CRUDL response shapes
 * @internal @typeParam IdentifierMeta - Resolve the identifier fields for all models
 * @internal @typeParam SecondaryIndexes - Map of model secondary index metadata
 */
type InternalClientSchemaMetadata<
  Schema extends GenericModelSchema<any>,
  NonModelTypes extends NonModelTypesShape,
  ResolvedSchema,
  ImplicitModelsIdentifierMeta,
  ResolvedFields extends Record<string, unknown>,
  IdentifierMeta extends Record<string, any> = ModelIdentifier<
    SchemaTypes<Schema>
  >,
  SecondaryIndexes extends Record<string, any> = Schema extends RDSModelSchema<
    any,
    any
  >
    ? object
    : ModelSecondaryIndexes<SchemaTypes<Schema>>,
> = {
  [__modelMeta__]: IdentifierMeta &
    ImplicitModelsIdentifierMeta &
    SecondaryIndexes &
    RelationalMetadata<ResolvedSchema, ResolvedFields, IdentifierMeta> &
    NonModelTypes &
    ResolveCustomOperations<Schema, ResolvedFields, NonModelTypes>;
};
