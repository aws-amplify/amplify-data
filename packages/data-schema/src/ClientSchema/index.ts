import type { __modelMeta__ } from '../runtime/client';
import type {
  BaseSchema,
  CustomOperation,
  CustomType,
  EnumType,
  GenericModelSchema,
  ModelSchemaContents,
  ModelType,
  RDSModelSchema,
} from '../ModelSchema';

import type {
  ClientCustomOperation,
  ClientCustomType,
  ClientEnum,
  ClientModel,
} from './Core';
import type {
  CombinedModelSchema,
  CombinedSchemaIndexesUnion,
} from '../CombineSchema';
import type { Brand, SpreadTuple } from '../util';

import { a } from '../index';

const schema = a.schema({
  Status: a.enum(['a', 'b', 'c']),

  getStatus: a
    .query()
    .arguments({
      postId: a.id(),
    })
    .returns(a.ref('Status')),

  echo: a
    .query()
    .arguments({
      message: a.string(),
    })
    .returns(a.string()),

  LikeResult: a.customType({
    likes: a.integer().required(),
  }),

  likePost: a
    .mutation()
    .arguments({
      postId: a.string().required(),
    })
    .returns(a.ref('LikeResult'))
    .handler(a.handler.function('likePost'))
    .authorization((allow) => allow.publicApiKey()),

  likeAllPosts: a
    .mutation()
    .returns(a.ref('LikeResult').array())
    .handler(a.handler.function('likeAllPosts'))
    .authorization((allow) => allow.publicApiKey()),

  Post: a
    .model({
      title: a.string().required(),
      description: a.string(),
      viewCount: a.integer(),
      updatedAt: a.string(),
      comments: a.hasMany('Comment', 'postId'),
    })
    .secondaryIndexes((index) => [
      index('title'),
      index('description')
        .queryField('myCustomIdx')
        .sortKeys(['updatedAt', 'viewCount']),
    ])
    .authorization((allow) => [allow.publicApiKey()]),

  Comment: a
    .model({
      body: a.string().required(),
      postId: a.id(),
      post: a.belongsTo('Post', 'postId'),
    })
    .authorization((allow) => [allow.publicApiKey()]),
});

export type ClientSchema<
  Schema extends GenericModelSchema<any> | CombinedModelSchema<any>,
> =
  Schema extends GenericModelSchema<any>
    ? InternalClientSchema<Schema>
    : Schema extends CombinedModelSchema<any>
      ? InternalCombinedSchema<Schema>
      : never;

type InternalClientSchema<
  CustomerSchema extends ModelSchemaContents | BaseSchema<any, any>,
> = CustomerSchema extends ModelSchemaContents
  ? {
      [K in keyof CustomerSchema]: ClientSchemaProperty<CustomerSchema, K>;
    }
  : InternalClientSchema<CustomerSchema['data']['types']>;

type ClientSchemaProperty<T extends ModelSchemaContents, K extends keyof T> =
  T[K] extends Brand<'enum'>
    ? RemapEnum<T, T[K]>
    : T[K] extends Brand<'customType'>
      ? RemapCustomType<T, T[K]>
      : T[K] extends Brand<
            | 'queryCustomOperation'
            | 'mutationCustomOperation'
            | 'subscriptionCustomOperation'
          >
        ? RemapCustomOperation<T, T[K]>
        : T[K] extends Brand<'modelType'>
          ? RemapModel<T, T[K]>
          : never;

type RemapEnum<_T extends ModelSchemaContents, E> =
  E extends EnumType<infer ET> ? ClientEnum<ET> : 'a';

type RemapCustomType<T extends ModelSchemaContents, E> =
  E extends CustomType<infer CT>
    ? ClientCustomType<InternalClientSchema<T>, CT>
    : 'b';

type RemapCustomOperation<T extends ModelSchemaContents, E> =
  E extends CustomOperation<infer CO, any>
    ? ClientCustomOperation<InternalClientSchema<T>, CO>
    : 'c';

type RemapModel<T extends ModelSchemaContents, E> =
  E extends ModelType<infer MT, any>
    ? ClientModel<InternalClientSchema<T>, MT>
    : 'd';

type _Schema = InternalClientSchema<typeof schema>;
type _T = _Schema['Post']['type'];

type GetInternalClientSchema<Schema> =
  Schema extends GenericModelSchema<any> ? InternalClientSchema<Schema> : never;

type CombinedClientSchemas<
  Schemas extends CombinedModelSchema<any>['schemas'],
> = {
  [Index in keyof Schemas]: Index extends CombinedSchemaIndexesUnion
    ? GetInternalClientSchema<Schemas[Index]>
    : never;
};

/**
 * Types for unwrapping and combining generic type args into client-consumable types
 * for multiple schemas
 *
 * @typeParam Combined - A container of multiple schemas
 *
 * @internal @typeParam ClientSchemas - The tuple of client schemas to combine
 */
type InternalCombinedSchema<
  Combined extends CombinedModelSchema<any>,
  ClientSchemas extends [...any] = CombinedClientSchemas<Combined['schemas']>,
> = SpreadTuple<{
  [I in keyof ClientSchemas]: I extends CombinedSchemaIndexesUnion
    ? Omit<ClientSchemas[I], typeof __modelMeta__>
    : never;
}>;
