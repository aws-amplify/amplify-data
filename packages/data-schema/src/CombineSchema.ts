import { DerivedApiDefinition } from '@aws-amplify/data-schema-types';
import { BaseSchema, GenericModelSchema } from './ModelSchema';
import { brand, Brand, IndexLimitUnion } from './util';

const COMBINED_SCHEMA_LIMIT = 50;

export type CombinedSchemaIndexesUnion = IndexLimitUnion<
  typeof COMBINED_SCHEMA_LIMIT
>[number];

const CombinedSchemaBrandName = 'CombinedSchema';
export const combinedSchemaBrand = brand(CombinedSchemaBrandName);
export type CombinedSchemaBrand = Brand<typeof CombinedSchemaBrandName>;

export type CombinedModelSchema<Schemas extends GenericModelSchema<any>[]> =
  CombinedSchemaBrand & { schemas: [...Schemas] } & BaseSchema<any>;

export function combine<Schema extends GenericModelSchema<any>[]>(
  schemas: [...Schema],
): CombinedModelSchema<Schema> {
  return internalCombine(schemas);
}

function internalCombine<
  Schema extends GenericModelSchema<any>[],
  SchemasTuple extends [...Schema],
>(schemas: SchemasTuple): CombinedModelSchema<Schema> {
  return {
    schemas: schemas,
    data: {
      types: schemas.reduce(
        (prev, schema) => ({ ...prev, ...schema.data.types }),
        {},
      ),
    },
    models: schemas.reduce(
      (prev, schema) => ({ ...prev, ...schema.models }),
      {},
    ),
    transform(): DerivedApiDefinition {
      const baseDefinition = {
        functionSlots: [],
        jsFunctions: [],
        schema: '',
        functionSchemaAccess: [],
        lambdaFunctions: {},
      };
      return schemas.reduce<DerivedApiDefinition>((prev, schema) => {
        const transformedSchema = schema.transform();
        return {
          functionSlots: [
            ...prev.functionSlots,
            ...transformedSchema.functionSlots,
          ],
          jsFunctions: [...prev.jsFunctions, ...transformedSchema.jsFunctions],
          schema: [prev.schema, transformedSchema.schema].join('\n'),
          functionSchemaAccess: [
            ...prev.functionSchemaAccess,
            ...transformedSchema.functionSchemaAccess,
          ],
          lambdaFunctions: {
            ...prev.lambdaFunctions,
            ...transformedSchema.lambdaFunctions,
          },
        };
      }, baseDefinition);
    },
    ...combinedSchemaBrand,
  };
}
