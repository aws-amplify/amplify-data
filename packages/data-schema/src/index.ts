import { schema } from './ModelSchema';
import { model } from './ModelType';
import {
  id,
  string,
  integer,
  float,
  boolean,
  date,
  time,
  datetime,
  timestamp,
  email,
  json,
  phone,
  url,
  ipAddress,
} from './ModelField';
import { ref } from './RefType';
import { hasOne, hasMany, belongsTo, manyToMany } from './ModelRelationalField';
import { allow } from './Authorization';
import { customType } from './CustomType';
import { enumType } from './EnumType';
import { query, mutation, subscription } from './CustomOperation';
import { handler } from './Handler';

// We are re-exporting with this pattern to make the JSDoc happy.
// Otherwise the JSDoc wouldn't consistently show up as a customer types
// "a."
export {
  schema,
  model,
  ref,
  customType,
  enumType as enum,
  query,
  mutation,
  subscription,
  hasOne,
  hasMany,
  belongsTo,
  manyToMany,
  allow,
  id,
  string,
  integer,
  float,
  boolean,
  date,
  time,
  datetime,
  timestamp,
  email,
  json,
  phone,
  url,
  ipAddress,
  handler,
};
