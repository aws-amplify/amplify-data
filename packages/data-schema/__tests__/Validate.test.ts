import { expectTypeTestsToPassAsync } from 'jest-tsd';
import { a } from '../src/index';

// evaluates type defs in corresponding test-d.ts file
it('should not produce static type errors', async () => {
  await expectTypeTestsToPassAsync(__filename);
});

describe('Validation directive generation', () => {
  test('Model with various validation rules without error messages', () => {
    const schema = a
      .schema({
        TestModel: a.model({
          id: a.id(),
          // Integer field with numeric validations
          integerField: a.integer().validate(v => 
            v.gt(0)
             .lt(100)
             .gte(1)
             .lte(99)
          ),
          positiveIntegerField: a.integer().validate(v => 
            v.positive()
          ),
          negativeIntegerField: a.integer().validate(v => 
            v.negative()
          ),
          // Float field with numeric validations
          floatField: a.float().validate(v => 
            v.gt(0.99)
             .lt(1000.01)
             .gte(1.0)
             .lte(1000.0)
          ),
          positiveFloatField: a.float().validate(v => 
            v.positive()
          ),
          negativeFloatField: a.float().validate(v => 
            v.negative()
          ),
          // String field with string validations
          stringField: a.string().validate(v => 
            v.minLength(5)
             .maxLength(20)
             .startsWith("ABC")
             .endsWith("XYZ")
             .matches("^[A-Z0-9]+$")
          ),
        }),
      })
      .authorization((allow) => allow.publicApiKey());

    expect(schema.transform().schema).toMatchSnapshot();
  });
  
  test('Model with various validation rules with error messages', () => {
    const schema = a
      .schema({
        TestModelWithErrors: a.model({
          id: a.id(),
          // Integer field with numeric validations
          integerField: a.integer().validate(v => 
            v.gt(0, "Value must be greater than 0")
             .lt(100, "Value must be less than 100")
             .gte(1, "Value must be at least 1")
             .lte(99, "Value must be at most 99")
          ),
          positiveIntegerField: a.integer().validate(v => 
            v.positive("Value must be positive")
          ),
          negativeIntegerField: a.integer().validate(v => 
            v.negative("Value must be negative")
          ),
          // Float field with numeric validations
          floatField: a.float().validate(v => 
            v.gt(0.99, "Value must be greater than 0.99")
             .lt(1000.01, "Value must be less than 1000.01")
             .gte(1.0, "Value must be at least 1.0")
             .lte(1000.0, "Value must be at most 1000.0")
          ),
          positiveFloatField: a.float().validate(v => 
            v.positive("Value must be positive")
          ),
          negativeFloatField: a.float().validate(v => 
            v.negative("Value must be negative")
          ),
          // String field with string validations
          stringField: a.string().validate(v => 
            v.minLength(5, "String must be at least 5 characters")
             .maxLength(20, "String must be at most 20 characters")
             .startsWith("ABC", "String must start with ABC")
             .endsWith("XYZ", "String must end with XYZ")
             .matches("^[A-Z0-9]+$", "String must only contain uppercase letters and numbers")
          ),
        }),
      })
      .authorization((allow) => allow.publicApiKey());

    expect(schema.transform().schema).toMatchSnapshot();
  });
});
