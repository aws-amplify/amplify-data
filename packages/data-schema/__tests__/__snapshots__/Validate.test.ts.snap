// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Validation directive generation Model with various validation rules with error messages 1`] = `
"type TestModelWithErrors @model @auth(rules: [{allow: public, provider: apiKey}])
{
  id: ID! @primaryKey
  integerField: Int @validate(type: gt, value: "0", errorMessage: ""Value must be greater than 0"") @validate(type: lt, value: "100", errorMessage: ""Value must be less than 100"") @validate(type: gte, value: "1", errorMessage: ""Value must be at least 1"") @validate(type: lte, value: "99", errorMessage: ""Value must be at most 99"")
  positiveIntegerField: Int @validate(type: gt, value: "0", errorMessage: ""Value must be positive"")
  negativeIntegerField: Int @validate(type: lt, value: "0", errorMessage: ""Value must be negative"")
  floatField: Float @validate(type: gt, value: "0.99", errorMessage: ""Value must be greater than 0.99"") @validate(type: lt, value: "1000.01", errorMessage: ""Value must be less than 1000.01"") @validate(type: gte, value: "1", errorMessage: ""Value must be at least 1.0"") @validate(type: lte, value: "1000", errorMessage: ""Value must be at most 1000.0"")
  positiveFloatField: Float @validate(type: gt, value: "0", errorMessage: ""Value must be positive"")
  negativeFloatField: Float @validate(type: lt, value: "0", errorMessage: ""Value must be negative"")
  stringField: String @validate(type: minLength, value: "5", errorMessage: ""String must be at least 5 characters"") @validate(type: maxLength, value: "20", errorMessage: ""String must be at most 20 characters"") @validate(type: startsWith, value: "ABC", errorMessage: ""String must start with ABC"") @validate(type: endsWith, value: "XYZ", errorMessage: ""String must end with XYZ"") @validate(type: matches, value: "^[A-Z0-9]+$", errorMessage: ""String must only contain uppercase letters and numbers"")
}"
`;

exports[`Validation directive generation Model with various validation rules without error messages 1`] = `
"type TestModel @model @auth(rules: [{allow: public, provider: apiKey}])
{
  id: ID! @primaryKey
  integerField: Int @validate(type: gt, value: "0") @validate(type: lt, value: "100") @validate(type: gte, value: "1") @validate(type: lte, value: "99")
  positiveIntegerField: Int @validate(type: gt, value: "0")
  negativeIntegerField: Int @validate(type: lt, value: "0")
  floatField: Float @validate(type: gt, value: "0.99") @validate(type: lt, value: "1000.01") @validate(type: gte, value: "1") @validate(type: lte, value: "1000")
  positiveFloatField: Float @validate(type: gt, value: "0")
  negativeFloatField: Float @validate(type: lt, value: "0")
  stringField: String @validate(type: minLength, value: "5") @validate(type: maxLength, value: "20") @validate(type: startsWith, value: "ABC") @validate(type: endsWith, value: "XYZ") @validate(type: matches, value: "^[A-Z0-9]+$")
}"
`;
